crates/token/src/lib.rs

```diff
@@ -108,21 +101,114 @@ impl Transfer {
     BorshDeserialize,
     BorshDeserializer,
     BorshSchema,
+    Default,
     Hash,
     Eq,
     PartialOrd,
     Serialize,
     Deserialize,
 )]
-pub struct TransferData {
-    /// Source address will spend the tokens
-    pub source: Address,
-    /// Target address will receive the tokens
-    pub target: Address,
-    /// Token's address
-    pub token: Address,
-    /// The amount of tokens
-    pub amount: DenominatedAmount,
+pub struct Transfer {
+    /// Sources of this transfer
+    pub sources: BTreeMap<Account, DenominatedAmount>,
+    /// Targets of this transfer
+    pub targets: BTreeMap<Account, DenominatedAmount>,
+    /// Hash of tx section that contains the MASP transaction
+    pub shielded_section_hash: Option<Hash>,
+}
+
+impl Transfer {
+    /// Create a MASP transaction
+    pub fn masp(hash: Hash) -> Self {
+        Self {
+            shielded_section_hash: Some(hash),
+            ..Self::default()
+        }
+    }
+
+    /// Set the key to the given amount
+    fn set<K: Ord>(
+        map: &mut BTreeMap<K, DenominatedAmount>,
+        key: K,
+        val: DenominatedAmount,
+    ) {
+        if val.is_zero() {
+            // Zero entries do not need to be present
+            map.remove(&key);
+        } else {
+            map.insert(key, val);
+        }
+    }
+
+    /// Credit the given account
+    pub fn credit(
+        mut self,
+        owner: Address,
+        token: Address,
+        amount: DenominatedAmount,
+    ) -> Option<Self> {
+        let account = Account { owner, token };
+        let zero = DenominatedAmount::new(Amount::zero(), amount.denom());
+        let source_amount = *self.sources.get(&account).unwrap_or(&zero);
+        let target_amount = *self.targets.get(&account).unwrap_or(&zero);
+        // If this account is already a target, then reduce the target
+        if amount < target_amount {
+            // Account still gets net increase
+            Self::set(
+                &mut self.targets,
+                account,
+                target_amount.checked_sub(amount)?,
+            );
+        } else {
+            // Account now actually gets a net decrease
+            self.targets.remove(&account);
+            let new_amt = source_amount
+                .checked_add(amount.checked_sub(target_amount)?)?;
+            Self::set(&mut self.sources, account, new_amt);
+        }
+        Some(self)
+    }
+
+    /// Debit the given account
+    pub fn debit(
+        mut self,
+        owner: Address,
+        token: Address,
+        amount: DenominatedAmount,
+    ) -> Option<Self> {
+        let account = Account { owner, token };
+        let zero = DenominatedAmount::new(Amount::zero(), amount.denom());
+        let source_amount = *self.sources.get(&account).unwrap_or(&zero);
+        let target_amount = *self.targets.get(&account).unwrap_or(&zero);
+        // If this account is already a source, then reduce the source
+        if amount < source_amount {
```

Comment:
If we want to debit an account, shouldn't we increase its source in any case? Or reduce the target if available?

Reply:
Oops, my mistake. The credit/debit function names are flipped. Will fix.